test/integration/quota/quota_test.go

```diff
@@ -477,56 +475,101 @@ func TestQuotaLimitService(t *testing.T) {
 			},
 		},
 	}
+
 	waitForQuota(t, quota, clientset)
 
-	tests := []testCase{
-		{
-			description: "node port service should be created successfully",
-			svc:         newService("np-svc", v1.ServiceTypeNodePort, true),
-			success:     true,
-		},
-		{
-			description: "first LB type service that allocates node port should be created successfully",
-			svc:         newService("lb-svc-withnp1", v1.ServiceTypeLoadBalancer, true),
-			success:     true,
-		},
-		{
-			description: "second LB type service that allocates node port creation should fail as node port quota is exceeded",
-			svc:         newService("lb-svc-withnp2", v1.ServiceTypeLoadBalancer, true),
-			success:     false,
-		},
-		{
-			description: "first LB type service that doesn't allocates node port should be created successfully",
-			svc:         newService("lb-svc-wonp1", v1.ServiceTypeLoadBalancer, false),
-			success:     true,
-		},
-		{
-			description: "second LB type service that doesn't allocates node port creation should fail as loadbalancer quota is exceeded",
-			svc:         newService("lb-svc-wonp2", v1.ServiceTypeLoadBalancer, false),
-			success:     false,
-		},
-		{
-			description: "forth service creation should be successful",
-			svc:         newService("clusterip-svc1", v1.ServiceTypeClusterIP, false),
-			success:     true,
-		},
-		{
-			description: "fifth service creation should fail as service quota is exceeded",
-			svc:         newService("clusterip-svc2", v1.ServiceTypeClusterIP, false),
-			success:     false,
-		},
+	// Creating the first node port service should succeed
+	nodePortService := newService("np-svc", v1.ServiceTypeNodePort, true)
+	_, err = clientset.CoreV1().Services(ns.Name).Create(context.TODO(), nodePortService, metav1.CreateOptions{})
+	if err != nil {
+		t.Errorf("creating first node port Service should not have returned error: %v", err)
 	}
 
-	for _, test := range tests {
-		t.Log(test.description)
-		_, err := clientset.CoreV1().Services(ns.Name).Create(context.TODO(), test.svc, metav1.CreateOptions{})
-		if (err == nil) != test.success {
-			if err != nil {
-				t.Fatalf("Error creating test service: %v, svc: %+v", err, test.svc)
-			} else {
-				t.Fatalf("Expect service creation to fail, but service %s is created", test.svc.Name)
-			}
+	// Creating the first loadbalancer service should succeed
+	lbServiceWithNodePort1 := newService("lb-svc-withnp1", v1.ServiceTypeLoadBalancer, true)
+	_, err = clientset.CoreV1().Services(ns.Name).Create(context.TODO(), lbServiceWithNodePort1, metav1.CreateOptions{})
+	if err != nil {
+		t.Errorf("creating first loadbalancer Service should not have returned error: %v", err)
+	}
+
+	// add a delay for resource quota changes to propagate
+	time.Sleep(1 * time.Second)
```

Comment:
Is there a more deterministic way to check quota? Or, is it a bug if we can actually create a svc that exceeds the quota before the quota is propagated?

Reply:
I think this is expected since resource quota controller reconciles in the background.

But I do think it's problematic that you can by-pass resource quota limits by making resources quickly and racing with resource quota controller. But I'm curious to hear @deads2k's opinion on this too.

Reply:
can we do the same that we are doing for endpoints?
https://github.com/kubernetes/kubernetes/blob/7cd40e188567116fbb762da84c6ebf3d3fac5399/test/integration/endpoints/endpoints_test.go#L136-L145

we check the quota resource in a wait.poll loop to know when the quota controller was updated

Reply:
+1

We should wait until the status is propagated correctly.

[That said - the fact that status is propagated doesn't immediately mean that admission plugin already has that state. So there will still be a possibility for races.]


The races in ResourceQuota are problematic but is a known bug for a long time - couple years ago @derekwaynecarr had a proposal for it, but it never materialized.

Reply:
@andrewsykim I think we can replace time.Sleep by the same helper used in the e2e
```
// wait for resource quota status to show the expected used resources value
func waitForResourceQuota(c clientset.Interface, ns, quotaName string, used v1.ResourceList) error {
	return wait.Poll(framework.Poll, resourceQuotaTimeout, func() (bool, error) {
		resourceQuota, err := c.CoreV1().ResourceQuotas(ns).Get(context.TODO(), quotaName, metav1.GetOptions{})
		if err != nil {
			return false, err
		}
		// used may not yet be calculated
		if resourceQuota.Status.Used == nil {
			return false, nil
		}
		// verify that the quota shows the expected used resource values
		for k, v := range used {
			if actualValue, found := resourceQuota.Status.Used[k]; !found || (actualValue.Cmp(v) != 0) {
				framework.Logf("resource %s, expected %s, actual %s", k, v.String(), actualValue.String())
				return false, nil
			}
		}
		return true, nil
	})
}
```
we know the values in quotas we want in each step

Reply:
Thanks -- I think this test does wait for quota update in `waitForQuota` but it's only on the initial creation of the resource, will add checks on each step

Reply:
Updated to waitForResourceQuota and I added you as a co-author in the last commit, thanks!